Megaparsec/Megaparsec.lean

```diff
@@ -58,21 +60,37 @@ inductive Result (S E A: Type) [Stream S] where
 | ok (x : A)
 | err (e : ParseError S E)
 
-structure Reply (S E A : Type) [Stream S] where
+structure Reply (S E A: Type) [Stream S] where
   state    : State S E
   consumed : Bool
   result   : Result S E A
 
-structure ParsecT (E S: Type) (M: Type -> Type) (A: Type) [Stream S] where
+structure ParsecT (E S: Type) [Stream S] (M: Type -> Type) [Monad M] (A: Type) where
   unParser :
     (B : Type) -> (State S E) ->
+    -- Return A with State S E and Hints into M B
     (A -> State S E -> Hints (Stream.Token S) -> M B) -> -- Consumed-OK
+    -- Report errors with State into M B
     (ParseError S E -> State S E -> M B) ->              -- Consumed-Error
+    -- Return A with State S E and Hints into M B
     (A -> State S E -> Hints (Stream.Token S) -> M B) -> -- Empty-OK
+    -- Report errors with State into M B
     (ParseError S E -> State S E -> M B) ->              -- Empty-Error
     M B
 
-def pMap (E S: Type) [Stream S] (M: Type -> Type) (U V: Type) (f: U -> V) (x: ParsecT E S M U) : ParsecT E S M V  :=
+private def run_cok (E S: Type) [Stream S] (M: Type -> Type) [Monad M] (A: Type) (a: A) (s₁: State S E) (_h: Hints (Stream.Token S)): M (Reply S E A) :=
+  pure $ Reply.mk s₁ True (Result.ok a)
+private def run_cerr (E S: Type) [Stream S] (M: Type -> Type) [Monad M] (A: Type) (err: ParseError S E) (s₁: State S E): M (Reply S E A) :=
```

Comment:
```lean
def run_cerr {E S: Type} [Stream S] {M: Type -> Type} [Monad M] {A: Type} (err: ParseError S E) (s₁: State S E): M (Reply S E A)
```

will work to get rid of `E S M A` explicit args in `pMap`, but I don't know if it has other implications. We also might want to group type params and typeclass args, putting the Stream and Monad constraint after `A : Type`

Reply:
Agree with both in principle, strongly disagree in practice in this particular case.

 2. It's cool to have constraints in one place, but jumping hither and thither is certainly not. This function head reads well in plain English "takes type E, a stream S, a monad M and a type A", rather than "takes types E S M and A, where S is a Stream and M is a Monad".